apps/algorithm/search/src/server.py

```diff
@@ -65,35 +129,99 @@ async def search(
     limit: int = 20,
     minPrice: float = None,
     maxPrice: float = None,
-    status: str = "AVAILABLE",
-    searchType: str = "LISTINGS",
-    sort: str = "RELEVANCE",
+    status: Status = "AVAILABLE",
+    searchType: SearchType = "LISTINGS",
+    sort: Sort = "RELEVANCE",
 ):
-    # actual logic will go here
-
-    # This is a dummy response
-    return [
-        {
-            "listingID": "A23F29039B23",
-            "sellerID": "A23F29039B23",
-            "sellerName": "John Doe",
-            "title": "Used Calculus Textbook",
-            "description": "No wear and tear, drop-off available.",
-            "price": 50,
-            "dateCreated": "2024-05-23T15:30:00Z",
-            "imageUrl": "image URL for first Image",
-        },
-        {
-            "listingID": "A23F29039B24",
-            "sellerID": "A23F29039B24",
-            "sellerName": "Jane Doe",
-            "title": "Used Physics Textbook",
-            "description": "No wear and tear, drop-off available.",
-            "price": 40,
-            "dateCreated": "2024-05-23T15:30:00Z",
-            "imageUrl": "image URL for first Image",
-        },
-    ]
+    try:
+        validate_search_params(latitude, longitude, page, limit, minPrice, maxPrice)
+
+        INDEX = os.getenv("ES_INDEX", DEFAULT_INDEX)
+
+        sort_options = {
+            "RELEVANCE": "_score",
+            "PRICE_ASC": "price",
+            "PRICE_DESC": "price",
+            "LISTED_TIME_ASC": "dateCreated",
+            "LISTED_TIME_DESC": "dateCreated",
+            "DISTANCE_ASC": "_geo_distance",
+            "DISTANCE_DESC": "_geo_distance",
+        }
+
+        if searchType == "LISTINGS":
+            must_conditions = [
+                {"multi_match": {"query": query, "fields": ["title", "description"]}},
+                {"match": {"status": status}},
+            ]
+        elif searchType == "USERS":
+            must_conditions = [
+                {"match": {"sellerName": query}},
+                {"match": {"status": status}},
+            ]
+        else:
+            raise HTTPException(status_code=422, detail="Invalid searchType")
+
+        search_body: Dict[str, Any] = {
+            "from": (page - 1) * limit,
+            "size": limit,
+            "query": {"bool": {"must": must_conditions, "filter": []}},
+            "sort": [],
+        }
+
+        if minPrice is not None or maxPrice is not None:
+            price_range = {}
+            if minPrice is not None:
+                price_range["gte"] = minPrice
+            if maxPrice is not None:
+                price_range["lte"] = maxPrice
+            search_body["query"]["bool"]["filter"].append(
+                {"range": {"price": price_range}}
+            )
+
+        search_body["query"]["bool"]["filter"].append(
+            {
+                "geo_distance": {
+                    "distance": "5km",
```

Comment:
Should this be hard-coded or perhaps a constant?

Reply:
Made it a constant for easy changing if we want to later